File: week3_greedy_algorithms/2_maximum_value_of_the_loot/fractional_knapsack.py
from sys import stdin

# Auxillary function to choose the best item every time
def BestItem (weights,values):
    max_wv = 0
    index = 0
    for i in range(len(weights)):
        if weights[i]>0:
            if values[i]/weights[i] > max_wv:
                max_wv = values[i]/weights[i]
                index = i 
    return index

#Greedy Way.
def optimal_value(capacity, weights, values):
    value = 0.0
    i = 0
    # write your code here
    
    while capacity > 0:
        i = BestItem(weights,values)
        a = min(capacity,weights[i])
        value += a * (values[i]/weights[i]) 
        weights[i] -= a
        capacity -= a    
    return value 


if __name__ == "__main__":
    # data = list(map(int, stdin.read().split()))
    data = [5, 5, 800, 1, 500, 1, 300, 2, 80, 5, 20, 6]
    n, capacity = data[0:2]
    values = data[2:(2 * n + 2):2]
    weights = data[3:(2 * n + 2):2]
    print(f""" {n}\n
{capacity}\n
{values}\n
{weights}\n 
""")
    opt_value = optimal_value(capacity, weights, values)
    print("{:.10f}".format(opt_value))

File: week3_greedy_algorithms/2_maximum_value_of_the_loot/fractional_knapsack.cpp
#include <iostream>
#include <vector>

using std::vector;


int BestItem(vector<int> weights, vector<int> values) {

  int max_wv = 0;
  int index = 0;

  for(int i=0; i<sizeof(weights); ++i)
  {
    if(weights[i]>0)
      if ((values[i]/weights[i]) > max_wv)
      {
        max_wv = values[i]/weights[i];
        index = i;
      }  
  }
  return index; 

}


double get_optimal_value(int capacity, vector<int> weights, vector<int> values) {
  double value = 0.0;
  int i = 0;
  int a = 0;

  while(capacity > 0)
  {
    i = BestItem(weights,values);
    a = std::min(capacity,weights[i]);
    value += a * (values[i]/weights[i]);
    weights[i] -= a;
    capacity -=a;
  }
  return value;

}

int main() {
  int n;
  int capacity;
  // std::cout<<"Enter items number and capacity"<<std::endl;
  std::cin >> n >> capacity;
  vector<int> values(n);
  vector<int> weights(n);
  for (int i = 0; i < n; i++) {
    // std::cout<<"Enter value and weight for item:"<<i<<std::endl;
    std::cin >> values[i] >> weights[i];
  }

  double optimal_value = get_optimal_value(capacity, weights, values);

  std::cout.precision(10);
  std::cout << optimal_value << std::endl;
  return 0;
}
